doc/stdenv/platform-notes.chapter.md

```diff
@@ -46,22 +46,189 @@ Some common issues when packaging software for Darwin:
+  | 14.0–14.3.1        | 13.3                                              | `apple-sdk_13`    |
+  | 15.0–15.4          | 14.4                                              | `apple-sdk_14`    |
+  | 16.0               | 15.0                                              | `apple-sdk_15`    |
+
+  To use a non-default SDK, add it to your build inputs.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "libfoo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 ]; # Uses the 15.0 SDK instead of the default SDK for the platform.
+  }
+  ```
+
+  If your derivation has multiple SDKs its inputs (e.g., because it has been propagated by one your dependencies), it will use the highest SDK version available.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "libfoo-1.2.3"; # Upstream specifies that it needs Xcode 12 to build, so use the 11.3 SDK.
+    # ...
+    buildInputs = [ apple-sdk_11 ];
+    nativeBuildInputs = [ swift ]; # Propagates the 13.3 SDK, so the 13.3 SDK package will be used instead of the 11.3 SDK.
+  }
+  ```
+
+- You should rely on the default SDK when possible. If a package specifies a required SDK version, use that version (e.g., libuv requires 11.0, so it should use `apple-sdk_11`). When a package supports multiple SDKs, determine which SDK package to use based on the following rules of thumb:
+
+  - If a package supports multiple SDK versions, use the lowest supported SDK version by the package (but no lower than the default SDK). That ensures maximal platform compatibility for the package.
+
+  - If a package specifies a range of supported SDK versions _and_ a minimum supported version, assume the package is using availability checks to support the indicated minimum version.Add latest supported SDK package and a `darwinMinVersionHook` set to the minimum supported by the upstream package.
+
+  Warning: Avoid using newer SDKs than an upstream package supports. When a binary is linked on Darwin, the SDK version used to build it is recorded in the binary. Runtime behavior can vary based on the SDK version, which may work fine but can also result in unexpected behavior or crashes when building with an unsupported SDK.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "foo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 (darwinMinVersionHook "10.15") ]; # Upstream builds with the 15.0 SDK but supports 10.15.
+  }
+  ```
+
+- Libraries that require a minimum version can propagate an appropriate SDK and `darwinMinVersionHook`. Derivations using that library will automatically use an appropriate SDK and minimum version. Even if the library builds with a newer SDK, it should propagate the minimum supported SDK. Derivations that need a newer SDK can add it to their `buildInputs`.
+
+  ```nix
+  stdenv.mkDerivation {
+    name = "libfoo-1.2.3";
+    # ...
+    buildInputs = [ apple-sdk_15 ]; # Upstream builds with the 15.0 SDK but supports 10.15.
```

Comment:
This is attempting to provide a simplified example of what Qt is doing. Qt builds with the 15.0 SDK, but it propagates an older one. The reason for that Darwin uses the linked SDK version to change runtime behavior.

A user-visible feature that is based on linked SDK is automatic dark mode. If your application is linked against the 10.14 SDK, Apple assumes that you did the work to support both light and dark modes. If you didn’t and link it anyway, then your application may look a bit funky or broken depending on the user’s settings

https://developer.apple.com/documentation/appkit/nsappearancecustomization/choosing_a_specific_appearance_for_your_macos_app

> The system assumes that apps linked against the macOS 10.14 or later SDK support both light and dark appearances. You can assign a specific appearance to windows, views, and popovers. You can also disable support for Dark Mode entirely using an `Info.plist` key.

A change in macOS 15 could potentially (if the application is accessing the registers directly instead of via recommended APIs) change application behavior. Certain registers return different values now on M3 and newer chips.

https://developer.apple.com/documentation/macos-release-notes/macos-15-release-notes#Platform

> On Apple Silicon based devices with M3 or later, and A16 Bionic or later, the values returned by reading the `CNTFRQ_EL0` and `CNTVCT_EL0` registers have been updated to 1 GHz, instead of the prior value of 24 MHz. It is still recommended for apps to use libsystem APIs like `mach_absolute_time()` for timekeeping. Your app will not be impacted by this change if it uses Apple’s timekeeping APIs. For compatibility purposes, this change will only be visible when using the SDK associated with this release or later. On macOS, applications running inside a Virtualization.framework VM will continue to receive the legacy behavior.

There have been other changes as well (such as to how certain widgets are rendered or laid out depending on the SDK version), but hopefully those examples help explain the reason for being conservative about SDK version.